subprojects/launcher/src/main/java/org/gradle/tooling/internal/provider/ContinuousBuildTriggerHandler.java

```diff
@@ -0,0 +1,279 @@
+/*
+ * Copyright 2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.tooling.internal.provider;
+
+import org.gradle.api.execution.internal.FileChangeListener;
+import org.gradle.api.execution.internal.TaskInputsListener;
+import org.gradle.api.internal.TaskInternal;
+import org.gradle.api.internal.file.FileCollectionInternal;
+import org.gradle.api.internal.file.FileCollectionStructureVisitor;
+import org.gradle.api.internal.file.FileTreeInternal;
+import org.gradle.api.internal.file.collections.FileSystemMirroringFileTree;
+import org.gradle.api.tasks.util.PatternSet;
+import org.gradle.execution.plan.BuildInputHierarchy;
+import org.gradle.initialization.BuildCancellationToken;
+import org.gradle.initialization.ContinuousExecutionGate;
+import org.gradle.internal.UncheckedException;
+import org.gradle.internal.filewatch.PendingChangesListener;
+import org.gradle.internal.logging.text.StyledTextOutput;
+import org.gradle.internal.os.OperatingSystem;
+import org.gradle.internal.watch.registry.FileWatcherRegistry;
+
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.gradle.internal.filewatch.DefaultFileSystemChangeWaiterFactory.QUIET_PERIOD_SYSPROP;
+
+public class ContinuousBuildTriggerHandler implements FileChangeListener, TaskInputsListener {
```

Comment:
Are we conflating too many things here? This is both two listeners, and handles the wait loop by itself.

Reply:
This class has been split up.